### PR TITLE
[asset backfill] RFC: Handle run success w/ no materialization

### PR DESCRIPTION
**Pull request by clairelin135**
Explores implementation for resolving https://github.com/dagster-io/dagster/issues/17745

**Problem statement**
When an asset backfill launches a run that completes successfully but does not materialize the asset (i.e. via optional output), the partition is stuck in an 'in progress' state forever, causing the backfill to hang indefinitely.

Impact: for certain cloud customers that use this pattern, hanging backfills cause network resource consumption (unsure of the details here) to grown unusually high.

**Background**
Because we don't store asset materialization failure events, in order to identify "unmaterialized but successful" partitions we have to query the backfill runs and use their statuses + asset materialization planned events to discover which partitions are done executing.

However, fetching runs can be expensive -- fetching all successful runs on each iteration would likely cause timeouts, so we need a bounded way to fetch runs launched by the backfill.

**Solution in this PR: In each iteration, fetch runs that completed since the prior iteration**
- We already store a `latest_storage_id` cursor that represents the latest event evaluated by the backfill
- In each iteration, fetch run success events after this cursor. Fetch the associated runs and filter by backfill ID to fetch just the backfill runs that have completed since the last iteration
- Use these runs to calculate which partitions are "unmaterialized but successfully executed".

Implementation details:
- `get_event_records` requires a run-sharded event cursor to fetch non-asset events (i.e. run success events) past a cursor in SQLite
    - Run success events are copied into the index shard, but unfortunately we can't change the behavior of `get_event_records` to search the index shard in this case since this would break existing run status sensors
    - This PR passes a `force_use_index_conn` param to force fetching run status events from the index shard. Realistically though I would prefer to store a `latest_evaluated_event_timestamp` value in `AssetBackfillData`  to build a run-sharded cursor

**Open questions**
- Should we mark these "unmaterialized but successful" partitions as successes or failures?
    - IMO these partitions are "successful" because they executed successfully
    - However, their downstreams need to be marked as failures since the upstream did not have a materialization (not implemented in this PR)
- Should we even allow this behavior? How high priority is fixing this issue?
    - Perhaps we should just find a way to stop backfills from hanging forever. I.e. in asset backfill iteration, if no runs have been launched in the last 24h but certain partitions are still hanging, we just mark the backfill as failed

### CODE REVIEW

**Comment by clairelin135** [comment]
Current dependencies on/for this PR:
* master
  * **PR #17748** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17748" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/17748?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 1]
```diff
@@ -1057,6 +1060,62 @@ def get_asset_backfill_iteration_materialized_partitions(
...
+            ),
+            force_use_index_connection=True,
+        )
+    ]
+
+    yield None
+
+    backfill_run_ids = []
+
+    for run_id in successful_run_ids:
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:1086]: there's opportunity to consolidate this with the logic in https://github.com/dagster-io/dagster/blob/57df2a919ce3b428cd03665c83129ae8febafa19/python_modules/dagster/dagster/_core/execution/asset_backfill.py#L1292-L1331

**Comment by sryza** [comment]
This is a great writeup.

> IMO these partitions are "successful" because they executed successfully. However, their downstreams need to be marked as failures since the upstream did not have a materialization (not implemented in this PR)

Marking their downstreams as failed seems undesirable to me in this case. It makes it seem like something went wrong, when, at least according to our run semantics, this is totally legit.

> (not implemented in this PR)

Does that mean there could still be a hanging backfill if there's an asset in the backfill that's downstream of the optional asset?

> Because we don't store asset materialization failure events

This makes me wonder whether we should bite the bullet and add these if it means we could avoid piling on more complexity?

**Comment by clairelin135** [comment]
> Marking their downstreams as failed seems undesirable to me in this case. It makes it seem like something went wrong, when, at least according to our run semantics, this is totally legit.

This sounds reasonable. We could fetch downstreams via `bfs_filter_subsets` or something and mark them all as successful even though they didn't execute.

This brings to mind another open question about behavior -- say a downstream partition C has two upstreams: optional output partition A and regular partition B. Should we execute C if A is skipped and B succeeds? IMO we should not execute. We currently have behavior where if any upstream partition fails, we fail. So following the same pattern, if any upstream skips, we should skip.

It also feels weird to add these partitions to a "materialized" subset. For full correctness, it might be desirable to add a "skipped" subset.

> Does that mean there could still be a hanging backfill if there's an asset in the backfill that's downstream of the optional asset?

Yep, that's right. 

> This makes me wonder whether we should bite the bullet and add these if it means we could avoid piling on more complexity?

In this case we would need `ASSET_MATERIALIZATION_SKIPPED`. This seems like a fairly narrow use case so for that reason not sure if it's worth supporting such a niche event type. Part of me wonders if it's more high value to just support a general functionality that discovers hanging backfills and marks them as failed. 

But I support `ASSET_MATERIALIZATION_FAILED`. The fact that we can't query for these forces us to unpack all of the failed/canceled backfill runs in each iteration, which probably has only succeeded performance-wise because there aren't that many failed runs. If someone were to kick off a backfill with 10K+ runs that all fail, I think we could certainly see daemon crashes with the currently implementation.

**Comment by sryza** [comment]
> This brings to mind another open question about behavior -- say a downstream partition C has two upstreams: optional output partition A and regular partition B. Should we execute C if A is skipped and B succeeds? IMO we should not execute. We currently have behavior where if any upstream partition fails, we fail. So following the same pattern, if any upstream skips, we should skip.

I agree. I believe this is the behavior within a single run, so we should mirror it within a backfill.

> Part of me wonders if it's more high value to just support a general functionality that discovers hanging backfills and marks them as failed.

For the situation outlined here, this doesn't seem like a great end state, because the backfill is actually doing what it's supposed to do successfully, right?  If this is a burning fire and we want to do the minimal-effort thing that allows us to focus on other priorities, then I could see this being an OK direction?

> In this case we would need ASSET_MATERIALIZATION_SKIPPED. This seems like a fairly narrow use case so for that reason not sure if it's worth supporting such a niche event type.

When you say "worth it", are you thinking about the cost in terms of performance or in terms of implementation / maintenance complexity? FWIW we have STEP_SKIPPED events.

**Comment by clairelin135** [comment]
> If this is a burning fire and we want to do the minimal-effort thing that allows us to focus on other priorities, then I could see this being an OK direction?

Yeah -- this is what I meant. I agree that it would be nice to support this use case, but only one org is using this pattern. So it kind of depends on what other priorities we have

> When you say "worth it", are you thinking about the cost in terms of performance or in terms of implementation / maintenance complexity? FWIW we have STEP_SKIPPED events.

Implementation/maintenance complexity. Good to know about the STEP_SKIPPED events, makes it more compelling to support MATERIALIZATION_SKIPPED events

**Comment by clairelin135** [comment]
Closing this PR as implementing ASSET_MATERIALIZATION_SKIPPED and ASSET_MATERIALIZATION_FAILURE events is a more performant way to solve this problem.

**Comment by cams-data** [comment]
@clairelin135 Sorry for the necro, the issue this PR was going to solve was never fixed.  AFAIK the events you mentioned in your final update have never been added to Dagster.  It is still an ongoing issue these years later, that if a single run backfill does not produce a materialize event, that the backfill daemon never detects completion.

**Pull request closed without merging**